Add experiment creation for CI

TerrenceMcGuinness-NOAA opened the pull request:
**Description**

Adding a third python script to create an experiment on the fly.
Also added detect_machine.sh.

Closes #1375

**Type of change**

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**
- [x] Cycled test on Hera with valid experiment dir in the PR staged area
  
**Checklist**
- [x] My code follows the style guidelines of this project

TerrenceMcGuinness-NOAA committed:
first few lines to get create_experment.py started

TerrenceMcGuinness-NOAA committed:
added first yaml config for creating experment cold_96

TerrenceMcGuinness-NOAA committed:
added executable from Executable and generated args for setup_expt from YAML

TerrenceMcGuinness-NOAA committed:
added commands to crate experment using Execute class

TerrenceMcGuinness-NOAA committed:
debugged first runthrough of create_experment.py

TerrenceMcGuinness-NOAA committed:
removed commentted out build_all and link

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1375_createexperment

TerrenceMcGuinness-NOAA committed:
improved yaml format, added Pathlib, and loggin

TerrenceMcGuinness-NOAA committed:
added detect_machine and defaluts if dected

TerrenceMcGuinness-NOAA committed:
got defalut to work and can use -h with detecte_manchine in use

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1375_createexperment

TerrenceMcGuinness-NOAA committed:
Update driver.sh

TerrenceMcGuinness-NOAA committed:
added just one BASH lint for echo consolidation

TerrenceMcGuinness-NOAA committed:
Update hera.sh

TerrenceMcGuinness-NOAA committed:
Update driver.sh

TerrenceMcGuinness-NOAA committed:
Update create_experment.py

TerrenceMcGuinness-NOAA committed:
debugging sciprts for makeing sure experment can run

TerrenceMcGuinness-NOAA committed:
couple of bash bugs for runthrough

TerrenceMcGuinness-NOAA committed:
added HOMEgfs into env that setup_expt.py runs

TerrenceMcGuinness-NOAA committed:
fixed bug intoduced by having HOMEgfs env se (created experment ran)

TerrenceMcGuinness-NOAA committed:
Update driver.sh

TerrenceMcGuinness-NOAA committed:
renamed second BASH script to clone-build_ci instead of run_ci

TerrenceMcGuinness-NOAA committed:
tightened code a little and added comments

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1375_createexperment

TerrenceMcGuinness-NOAA commented:
Automated global-workflow Testing Results:
Machine: Hera
```
Start: Mon Mar 13 16:56:46 UTC 2023 on hfe11
---------------------------------------------------
Build:                                 *SUCCESS*
Build: Completed at Mon Mar 13 17:26:10 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 18:06:31 UTC 2023

TerrenceMcGuinness-NOAA committed:
moved python install and added module load base in driver

TerrenceMcGuinness-NOAA committed:
fixed cp -as with wild card

TerrenceMcGuinness-NOAA commented:
Automated global-workflow Testing Results:
Machine: Hera
```
Start: Mon Mar 13 16:56:46 UTC 2023 on hfe11
---------------------------------------------------
Build:                                 *SUCCESS*
Build: Completed at Mon Mar 13 17:26:10 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 18:06:31 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 20:43:43 UTC 2023

TerrenceMcGuinness-NOAA commented:
Automated global-workflow Testing Results:
Machine: Hera
```
Start: Mon Mar 13 16:56:46 UTC 2023 on hfe11
---------------------------------------------------
Build:                                 *SUCCESS*
Build: Completed at Mon Mar 13 17:26:10 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 18:06:31 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 20:43:43 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 21:40:11 UTC 2023

TerrenceMcGuinness-NOAA commented:
Automated global-workflow Testing Results:
Machine: Hera
```
Start: Mon Mar 13 16:56:46 UTC 2023 on hfe11
---------------------------------------------------
Build:                                 *SUCCESS*
Build: Completed at Mon Mar 13 17:26:10 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 18:06:31 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 20:43:43 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 21:40:11 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 21:42:00 UTC 2023

TerrenceMcGuinness-NOAA commented:
Automated global-workflow Testing Results:
Machine: Hera
```
Start: Mon Mar 13 16:56:46 UTC 2023 on hfe11
---------------------------------------------------
Build:                                 *SUCCESS*
Build: Completed at Mon Mar 13 17:26:10 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 18:06:31 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 20:43:43 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 21:40:11 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 21:42:00 UTC 2023
Created experment
Experment setup: Completed at Mon Mar 13 21:45:42 UTC 2023

TerrenceMcGuinness-NOAA committed:
fixed bug with wild cards using subproess

WalterKolczynski-NOAA commented on ci/driver.sh:
```diff
@@ -99,21 +114,43 @@ for pr in ${pr_list}; do
+    module use "${pr_dir}"/global-workflow/modulefiles
+    module load "module_base.${TARGET}"
+    cd "${pr_dir}/global-workflow/ush/python/pygw"
+    python3 -m pip install .
```
Is it necessary to pip install?

TerrenceMcGuinness-NOAA replied:
```
Terry.McGuinness (hfe12) ~ $ work
Terry.McGuinness (hfe12) Terry.McGuinness $ cd global-workflow/ci
Terry.McGuinness (hfe12) ci (feature_1375_createexperment) $ export HOMEGFS=/scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1388_pass/global-workflow
Terry.McGuinness (hfe12) ci (feature_1375_createexperment) $ export RUNTEST=/scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1388_pass/RUNTEST
Terry.McGuinness (hfe12) ci (feature_1375_createexperment) $ ./create_experment.py --yaml cold_96_00z.yaml 
Traceback (most recent call last):
  File "./create_experment.py", line 16, in <module>
    from pygw.yaml_file import YAMLFile
ModuleNotFoundError: No module named 'pygw'
Terry.McGuinness (hfe12) ci (feature_1375_createexperment) $ cd ..
Terry.McGuinness (hfe12) global-workflow (feature_1375_createexperment) $ module use modulefiles
Terry.McGuinness (hfe12) global-workflow (feature_1375_createexperment) $ module load module_base.hera
(ufswm) Terry.McGuinness (hfe12) global-workflow (feature_1375_createexperment) $ cd ci
(ufswm) Terry.McGuinness (hfe12) ci (feature_1375_createexperment) $ ./create_experment.py --yaml cold_96_00z.yaml 
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/Terry.McGuinness/global-workflow/ci/./create_experment.py", line 22, in <module>
    from hosts import Host
ModuleNotFoundError: No module named 'hosts'
(ufswm) Terry.McGuinness (hfe12) ci (feature_1375_createexperment) $ export PYTHONPATH=/scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1388_pass/global-workflow/workflow
(ufswm) Terry.McGuinness (hfe12) ci (feature_1375_createexperment) $ ./create_experment.py --yaml cold_96_00z.yaml 
2023-03-14 16:47:58,634 - INFO     - root        : Running on HOST:HERA
2023-03-14 16:47:58,637 - INFO     - root        : Run command: /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1388_pass/global-workflow/workflow/setup_expt.py cycled --pslot cold_96_00z --app ATM --resdet 96 --resens 48 --comrot /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1388_pass/RUNTEST --expdir /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1388_pass/RUNTEST/expdir --idate 2021122018 --edate 2021122200 --nens 2 --gfs_cyc 1 --start cold
2023-03-14 16:47:58,906 - INFO     - root        : Run command: cp -as /scratch1/NCEPDEV/global/glopara/data/ICSDIR/C96C48/gdas.20211220 /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1388_pass/RUNTEST/cold_96_00z
2023-03-14 16:47:58,944 - INFO     - root        : Run command: cp -as /scratch1/NCEPDEV/global/glopara/data/ICSDIR/C96C48/enkfgdas.20211220 /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1388_pass/RUNTEST/cold_96_00z
2023-03-14 16:48:00,216 - INFO     - root        : Run command: /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1388_pass/global-workflow/workflow/setup_xml.py /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1388_pass/runtest/expdir/cold_96_00z
(ufswm) Terry.McGuinness (hfe12) ci (feature_1375_createexperment) $ 
```

TerrenceMcGuinness-NOAA replied:
Looks like you are correct, there was no need to run `pip install .`, but I did have to set **PYHTONPATH** to pick up the Host class in the workflow dir, maybe that should be in `ush/python/pygw` ?

WalterKolczynski-NOAA replied:
Talk to @aerorahul. He has been coordinating pythonpath stuff for the new UFSDA material.

TerrenceMcGuinness-NOAA replied:
As you know, It turns out that I still needed it to pick up YAML from **setup.cnf** since it was not in HPC-Stack.  Thanks for suggesting the same fix I was thinking (on Monday).

WalterKolczynski-NOAA commented on ci/create_experment.py:
```diff
@@ -0,0 +1,87 @@
+_top  = Path.absolute(Path(__file__)).parents[1]
+
+def input_args():
+    """
+    Method to collect user arguments for `create_experment.py`
+    """
+    description = """
+        Single agument as a yaml file
+        """
```
Please update to use numpy/reST format: https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard

TerrenceMcGuinness-NOAA replied:
Done .. @WalterKolczynski-NOAA double check these updates if this is what you meant.

WalterKolczynski-NOAA replied:
Nope. You don't have to repeat the function name, but you do need to flesh out the rest more. Collect them how? What is the return value? See both the numpy documentation I provided and https://github.com/NOAA-EMC/global-workflow/blob/develop/ush/python/pygfs/task/analysis.py

Also, `description` probably should just be a single-line string with a blank line separating it from the docblock.  I mistook it for part of the docblock the first time.

TerrenceMcGuinness-NOAA replied:
ok

WalterKolczynski-NOAA replied:
This docblock is still not in numpy format. Needs to be something like this:
```
"""
Method to collect user arguments
 
Parses all of the command-line arguments with argparse and returns a namespace with the values. Currently the only argument is the path to a YAML file that contains the experiment information. That YAML file should contain the following:

<list parameters of YAML with descriptions>

Returns
--------
args:   Namespace
    Namespace containing all of the command-line arguments
"""
```

WalterKolczynski-NOAA replied:
This one is still unresolved.

TerrenceMcGuinness-NOAA replied:
ok, still got to get that right, thanks.

TerrenceMcGuinness-NOAA replied:
So why was it not resolved, I had ''' instead of """?  I added a Return comment section incase that was the issue.

TerrenceMcGuinness-NOAA replied:
Ok, I've been asked to give more content on the workings of the YAML file itself.  I agree this needs some edification.

TerrenceMcGuinness-NOAA replied:
Ok, pushed the updates for the doc-block, I hope this cuts the mustered this time.

WalterKolczynski-NOAA commented on ci/create_experment.py:
```diff
@@ -0,0 +1,87 @@
+    logger.info(f'Run command: {setup_expt_cmd.command}')
+    setup_expt_cmd(output='stdout_expt', error='stderr_expt') 
+
+    # Make symlinks of initial conitions into ROTDIR
+    for folder in icdir:
+        link_ic_cmd = Executable('cp')
+        link_ic_cmd.add_default_arg('-as')
+        link_ic_cmd.add_default_arg(folder)
+        link_ic_cmd.add_default_arg(rotdir)
+        logger.info(f'Run command: {link_ic_cmd.command}')
+        link_ic_cmd()
```
I know you are emulating my own script here, but I'm pretty sure `setup_exp.py` will already stage the ICs for you with the `--icsdir` option.

TerrenceMcGuinness-NOAA replied:
No kidding!  Oh man, that could have saved me some time too, had to work the glob thing a little. I think we **_did_** have that before.

TerrenceMcGuinness-NOAA replied:
Done, a lot faster too!

WalterKolczynski-NOAA commented on ci/create_experment.py:
```diff
@@ -0,0 +1,87 @@
+        link_ic_cmd()
+
+    setup_xml_cmd = Executable(Path.absolute(Path.joinpath(Path(HOMEgfs),'workflow','setup_xml.py')))
+    expdir = Path.absolute(Path.joinpath(Path(setup_expt_args.arguments.expdir),Path(setup_expt_args.arguments.pslot)))
```
Is it necessary to create Path objects to feed into `Path.joinpath()`, or can you just use the strings and simplify this line?

TerrenceMcGuinness-NOAA replied:
Yes they are necessary ...
```
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/Terry.McGuinness/global-workflow/ci/./create_experment.py", line 82, in <module>
    expdir = Path.absolute(Path.joinpath(setup_expt_args.arguments.expdir,setup_expt_args.arguments.pslot))
  File "/scratch2/NCEPDEV/ensemble/save/Walter.Kolczynski/hpc-stack/core/miniconda3/4.6.14/envs/ufswm/lib/python3.9/pathlib.py", line 972, in joinpath
    return self._make_child(args)
AttributeError: 'str' object has no attribute '_make_child'
```

TerrenceMcGuinness-NOAA commented:
Automated global-workflow Testing Results:
Machine: Hera
```
Start: Tue Mar 14 15:09:59 UTC 2023 on hfe04
---------------------------------------------------
Build:                                 *SUCCESS*
Build: Completed at Tue Mar 14 15:38:48 UTC 2023
Created experment
Experment setup: Completed at Tue Mar 14 15:39:23 UTC 2023

TerrenceMcGuinness-NOAA committed:
Update driver.sh

TerrenceMcGuinness-NOAA committed:
Update create_experment.py

TerrenceMcGuinness-NOAA committed:
removed cp -as command and replaced it with the icsdir switch

WalterKolczynski-NOAA commented on ci/cold_96_00z.yaml:
```diff
@@ -0,0 +1,20 @@
+environment:
+  icdir: /scratch1/NCEPDEV/global/glopara/data/ICSDIR/C96C48
```
This path is going to be different on different machines. You likely should parameterize all but the last bit and add the definitions to ${machine}.sh

TerrenceMcGuinness-NOAA replied:
Oh yes, I have not generated these yet, I planned on doing that in another PR.

TerrenceMcGuinness-NOAA replied:
Oh yes, easy fix.  On another note, this moved because I updated yaml to use **icsdir** and forgot to check it in with the update.

TerrenceMcGuinness-NOAA committed:
forgot to git add yaml update for idsdir

TerrenceMcGuinness-NOAA committed:
updated rocoto_viwer in this branch to show the fix is replacing getitorator with iter

aerorahul commented on ci/cold_96_00z.yaml:
```diff
@@ -0,0 +1,20 @@
+   resens: 48
+   comrot: ${RUNTEST}
+   expdir: ${RUNTEST}/expdir
+   icsdir: /scratch1/NCEPDEV/global/glopara/data/ICSDIR/C96C48
```
should probably do something like:
```suggestion
   icsdir: ${ICSDIR_ROOT}/C96C48
```
and then define `ICSDIR_ROOT` in `${machine}.sh`
I think it will help with portability of the test.

TerrenceMcGuinness-NOAA replied:
Yes,  Walter noted this too, I get it now, I will update this next.

aerorahul commented on ci/hera.sh:
```diff
@@ -1,8 +1,9 @@
 export SLURM_QOS=debug
-export repo_url="https://github.com/NOAA-EMC/global-workflow.git"
\ No newline at end of file
+export repo_url="https://github.com/NOAA-EMC/global-workflow.git"
```
```suggestion
export ICSDIR_ROOT="/scratch1/NCEPDEV/global/glopara/data/ICSDIR"
```
to go with the comment in `cold_96_00z.yaml`

aerorahul commented on ci/driver.sh:
```diff
@@ -23,48 +23,62 @@ repo_url=${repo_url:-"https://github.com/NOAA-EMC/global-workflow.git"}
 ################################################################
-set -eux
-pwd="$(cd "$(dirname  "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd )"
+script_pwd="$(cd "$(dirname  "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd )"
```
```suggestion
ROOT_DIR="$(cd "$(dirname  "${BASH_SOURCE[0]}")/.." >/dev/null 2>&1 && pwd )"
```
Instead of `script_pwd` and then using relative paths, IMO, it is better to find the `ROOT_DIR` and then use that wherever needed.  So you don't have to guess where you are in the script.

aerorahul commented on ci/driver.sh:
```diff
@@ -23,48 +23,62 @@ repo_url=${repo_url:-"https://github.com/NOAA-EMC/global-workflow.git"}
-  esac
-done
+
+source "${script_pwd}/../ush/detect_machine.sh"
```
```suggestion
source "${ROOT_DIR}/ush/detect_machine.sh"
```
In light of the comment above.

aerorahul commented on ci/driver.sh:
```diff
@@ -23,48 +23,62 @@ repo_url=${repo_url:-"https://github.com/NOAA-EMC/global-workflow.git"}
   hera | orion)
     echo "Running Automated Testing on ${TARGET}"
-    source "${pwd}/${TARGET}.sh"
+    source "${script_pwd}/${TARGET}.sh"
```
```suggestion
    source "${ROOT_DIR}/ci/${TARGET}.sh"
```

aerorahul commented on ci/driver.sh:
```diff
@@ -99,21 +114,42 @@ for pr in ${pr_list}; do
-  "${pwd}/run_ci.sh" -p "${pr}" -d "${pr_dir}" -o "${pr_dir}/output_${id}"
+  # call clone-build_ci to clone and build PR
+  id=$("${GH}" pr view "${pr}" --repo "${repo_url}" --json id --jq '.id')
+  "${script_pwd}/clone-build_ci.sh" -p "${pr}" -d "${pr_dir}" -o "${pr_dir}/output_${id}"
```
```suggestion
  "${ROOT_DIR}/ci/clone-build_ci.sh" -p "${pr}" -d "${pr_dir}" -o "${pr_dir}/output_${id}"
```

aerorahul commented on ci/create_experment.py:
```diff
@@ -0,0 +1,82 @@
+'''
+
+import os, sys, socket
+import glob
```
I don't think you are using this anymore.
```suggestion
```

TerrenceMcGuinness-NOAA replied:
oh right.

aerorahul commented on ci/create_experment.py:
```diff
@@ -0,0 +1,82 @@
+from hosts import Host
+
+logger = Logger(level='DEBUG', colored_log=True)
+_here = Path.absolute(Path(__file__)).parents[0]
+_top  = Path.absolute(Path(__file__)).parents[1]
```
```suggestion
```
These are unused.

TerrenceMcGuinness-NOAA replied:
right again

aerorahul commented on ci/create_experment.py:
```diff
@@ -0,0 +1,82 @@
+    """
+    input_args()
+    
+    Method to collect user arguments for `create_experment.py`
```
```suggestion
    Method to collect user arguments for `create_experiment.py`
```

TerrenceMcGuinness-NOAA replied:
oops, still had one more, should have "grep-ed" for it

aerorahul commented on ci/create_experment.py:
```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env python3
```
typo in the filename. I think you meant`create_experiment.py`

aerorahul commented on ci/create_experment.py:
```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env python3
+
+'''Basic python script to create an experment directory on the fly from a given
```
```suggestion
'''Basic python script to create an experiment directory on the fly from a given
```

aerorahul commented on ci/cold_96_00z.yaml:
```diff
@@ -0,0 +1,20 @@
+  rotdir: ${RUNTEST}/cold_96_00z
+  HOMEgfs: ${HOMEGFS}   #TODO - currenlty using the env var ${HOMEgfs}
+                        #cause XML file to have HOMEgfs set to None 
+experment:
```
```suggestion
experiment:
```

aerorahul commented on ci/create_experment.py:
```diff
@@ -0,0 +1,82 @@
+    rotdir = str(Path.absolute(Path(setup_expt_args.environment.rotdir)))
+    HOMEgfs = setup_expt_args.environment.HOMEgfs
+
+    mode = setup_expt_args.experment.mode
```
```suggestion
    mode = setup_expt_args.experiment.mode
```

aerorahul commented on ci/driver.sh:
```diff
@@ -23,48 +23,62 @@ repo_url=${repo_url:-"https://github.com/NOAA-EMC/global-workflow.git"}
-done
+
+source "${script_pwd}/../ush/detect_machine.sh"
+if [[ "${MACHINE_ID}" != "UNKNOWN" ]]; then
+   TARGET="${MACHINE_ID}"
+fi   
```
Shouldn't there be an `exit 1` for the else part where `MACHINE_ID == UNKNOWN`?

TerrenceMcGuinness-NOAA replied:
It appears like that but actually, there's a bunch of logic after that so you can use the switch too, then it will exit if still not supported.

TerrenceMcGuinness-NOAA replied:
```
(ufswm) Terry.McGuinness (hfe12) ci (feature_1375_createexperment) $ ./driver.sh -h
Begin driver.sh at Tue Mar 14 21:51:01 UTC 2023

Usage: ./driver.sh -t <target> -h

  -h  display this message and quit
  -t  target/machine script is runningon DEFAULT: hera
```

TerrenceMcGuinness-NOAA replied:
notice it picks up the DEFAULT when using -h

aerorahul commented on ci/driver.sh:
```diff
@@ -23,48 +23,62 @@ repo_url=${repo_url:-"https://github.com/NOAA-EMC/global-workflow.git"}
+   TARGET="${MACHINE_ID}"
+fi   
+
+if [[ -z ${TARGET+x} || $1 == "-h" ]]; then
+  export TARGET
+  while getopts "t:h" opt; do
+    case ${opt} in
+      t)
+        TARGET=${OPTARG}
+        ;;
+      h|\?|:)
+        usage
+        ;;
+      *)
+        echo "Unrecognized option" 
+        usage
+    esac
+  done
+
+  if [[ -z "${TARGET+x}" ]]; then
+    usage
+  fi  
+
+fi
```
I don't think we need this and the `usage` method anymore since we are using `detect_machine.sh` to detect the `TARGET`.  This script no longer needs to be provided a `TARGET` as input when called.

TerrenceMcGuinness-NOAA replied:
Oh ok, I'll take it out then.

aerorahul reviewed (commented):
Looks good mostly.

I have some general comments in the review.

Specific comments:
- See comment on `ICSDIR_ROOT`
- See comment on `DIR_ROOT`
- spellcheck please.

TerrenceMcGuinness-NOAA committed:
replace all 100 mispellings of experiment, addressed PR comments

TerrenceMcGuinness-NOAA committed:
took out rocoto_viewer updates because alread in develop from hotfix PR

TerrenceMcGuinness-NOAA committed:
missed one more misspelling of experiment

TerrenceMcGuinness-NOAA commented:
Automated global-workflow Testing Results:
Machine: Hera
```
Start: Tue Mar 14 22:18:08 UTC 2023 on hfe06
---------------------------------------------------
Build:                                 *SUCCESS*
Build: Completed at Tue Mar 14 22:46:57 UTC 2023
Created experiment
Experiment setup: Completed at Tue Mar 14 23:04:50 UTC 2023

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1375_createexperment

TerrenceMcGuinness-NOAA committed:
created sub directories under {HOMEgfs}/ci and a setup module in {HOMEgfs} modulefiles for setup enviromnent

TerrenceMcGuinness-NOAA commented:
Automated global-workflow Testing Results:
Machine: Hera
```
Start: Tue Mar 14 22:18:08 UTC 2023 on hfe06
---------------------------------------------------
Build:                                 *SUCCESS*
Build: Completed at Tue Mar 14 22:46:57 UTC 2023
Created experiment
Experiment setup: Completed at Tue Mar 14 23:04:50 UTC 2023
Created experiment
Experiment setup: Completed at Thu Mar 16 17:22:47 UTC 2023

TerrenceMcGuinness-NOAA committed:
made changes for first run through debug

TerrenceMcGuinness-NOAA committed:
put call to build clone back in that was for debugging

TerrenceMcGuinness-NOAA commented:
Automated global-workflow Testing Results:
Machine: Hera
```
Start: Thu Mar 16 17:33:25 UTC 2023 on hfe08
---------------------------------------------------
Build:                                 *SUCCESS*
Build: Completed at Thu Mar 16 18:03:58 UTC 2023
Created experiment
Experiment setup: Completed at Thu Mar 16 18:04:06 UTC 2023

TerrenceMcGuinness-NOAA committed:
added PSLOT experment name to GitHub message that only had date before

aerorahul reviewed (approved):
Looks good.

TerrenceMcGuinness-NOAA committed:
doc blocks and removed vestigial imports to resolve PR comments

aerorahul commented:
please fix the failing test and we are good to merge this PR

aerorahul commented:
oops sorry.  closed accidentally instead of just comment.

TerrenceMcGuinness-NOAA committed:
fixing pynorms

TerrenceMcGuinness-NOAA committed:
fixing pycodestyls

TerrenceMcGuinness-NOAA committed:
fixing more pycodestyls

TerrenceMcGuinness-NOAA committed:
few more pycodestyles

TerrenceMcGuinness-NOAA committed:
pycodestyle comples about too many and too few blank lines

TerrenceMcGuinness-NOAA committed:
pycodestyle still complains about too many and too few blank lines

TerrenceMcGuinness-NOAA committed:
pycodestyle blank lines, think I finially get it

aerorahul commented on ci/scripts/driver.sh:
```diff
@@ -85,7 +82,7 @@ if [[ -s "${GFS_CI_ROOT}/${pr_list_file}" ]]; then
 else
  echo "no PRs to process .. exit"
- exit
+ exit 1
```
Why `exit 1` when there are no PR's to process?
Is that an error in the script?

WalterKolczynski-NOAA replied:
Agree, this seems like a success so it shouldn't return non-zero.

TerrenceMcGuinness-NOAA replied:
Oh sorry, right.

TerrenceMcGuinness-NOAA committed:
added BASH lines for multiple experiments

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_ci_multiple

TerrenceMcGuinness-NOAA committed:
cleaned up and debugged multiple experments loop in driver

WalterKolczynski-NOAA commented on ci/scripts/driver.sh:
```diff
@@ -23,62 +23,44 @@ repo_url=${repo_url:-"https://github.com/NOAA-EMC/global-workflow.git"}
+  echo "This is top level script to run CI tests on the global-workflow repo"
   if [[ -n "${TARGET+x}" ]]; then
-    echo "on DEFAULT: ${TARGET}"
+     echo "on the DEFAULTED: ${TARGET} machine"
```
"On the defaulted"?

WalterKolczynski-NOAA commented on modulefiles/module_setup.hera.lua:
```diff
@@ -0,0 +1,14 @@
+prepend_path("MODULEPATH", "/scratch2/NCEPDEV/ensemble/save/Walter.Kolczynski/hpc-stack/modulefiles/stack")
+load(pathJoin("hpc", "1.2.0"))
+load(pathJoin("hpc-miniconda3", "4.6.14"))
+load(pathJoin("ufswm", "1.0.0"))
```
Load `gfs_workflow` instead of `ufswm` and you won't have to `pip install yaml`.

TerrenceMcGuinness-NOAA replied:
oh right, now I get why this was here, makes perfect sense now, thanks.

WalterKolczynski-NOAA commented on ci/scripts/driver.sh:
```diff
@@ -116,27 +98,36 @@ for pr in ${pr_list}; do
+    if [[ ! -L pygw ]]; then
+      ln -s "${HOMEGFS}/ush/python/pygw/src/pygw" pygw
+    fi
+    PSLOT=C96C48_hybatmDA # TODO loop over experments for each yaml file in the experiments dir
```
Note on the TODO: We likely don't want to always run all of the defined experiments, so we'll want to have a list to test to run somewhere.

TerrenceMcGuinness-NOAA replied:
I just coded this today to run through all the yaml files in the experiments dir instead, sense you mention that here, I'm going to update the PR with that update, I've all already tested it.

TerrenceMcGuinness-NOAA committed:
few updates from PR comments (exit 0, gfs_workflow)

TerrenceMcGuinness-NOAA committed:
Update driver.sh

TerrenceMcGuinness-NOAA committed:
Update driver.sh

TerrenceMcGuinness-NOAA committed:
Update driver.sh

TerrenceMcGuinness-NOAA committed:
Update driver.sh

TerrenceMcGuinness-NOAA committed:
updated extention removal on experiment list

TerrenceMcGuinness-NOAA committed:
Update driver.sh

TerrenceMcGuinness-NOAA committed:
push file again to update GitHub actions

TerrenceMcGuinness-NOAA committed:
Update driver.sh

TerrenceMcGuinness-NOAA committed:
Update driver.sh

WalterKolczynski-NOAA commented on ci/scripts/driver.sh:
```diff
@@ -121,27 +118,42 @@ for pr in ${pr_list}; do
+      else 
+        {
+          echo "Failed on createing experiment ${pslot}"
+          echo "Experiment setup: failed at $(date) for expirment ${pslot}" || true
```
Experiment mispelled

TerrenceMcGuinness-NOAA replied:
now that was just simply lazy for me to miss that.

WalterKolczynski-NOAA commented on ci/scripts/driver.sh:
```diff
@@ -121,27 +118,42 @@ for pr in ${pr_list}; do
+          echo "Experiment setup: failed at $(date) for expirment ${pslot}" || true
+        } >> "${GFS_CI_ROOT}/PR/${pr}/output_${id}"
+        "${GH}" pr edit "${pr}" --repo "${repo_url}" --remove-label "${CI_HOST}-Running" --add-label "${CI_HOST}-Failed"
+        exit 1
```
Not sure if we want to exit here, as it will prevent the processing of additional PRs.

TerrenceMcGuinness-NOAA replied:
oh left that in from when there was one experiment, thanks!!

WalterKolczynski-NOAA commented on ci/scripts/driver.sh:
```diff
@@ -121,27 +118,42 @@ for pr in ${pr_list}; do
+    {
+      echo "Failed on cloning and building global-workflowi PR: ${pr}"
+      echo "CI on ${CI_HOST} failed to build on $(date) for repo ${repo_url}}" || true
+    } >> "${GFS_CI_ROOT}/PR/${pr}/output_${id}"
```
Doesn't `${GH}` need to be called here to label the PR as failed?

TerrenceMcGuinness-NOAA replied:
Of course!! Thanks again.

TerrenceMcGuinness-NOAA committed:
responded to some keen review comments on some key logic errors

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1375_createexperment

TerrenceMcGuinness-NOAA commented:
Automated global-workflow Testing Results:
Machine: Hera
```
Start: Wed Mar 22 17:32:30 UTC 2023 on hfe06
---------------------------------------------------
Build:                                 *SUCCESS*
Build: Completed at Wed Mar 22 18:01:15 UTC 2023
Created experiment
Experiment setup: Completed at Wed Mar 22 18:01:31 UTC 2023 for expirment C96C48_hybatmDA_also
Created experiment
Experiment setup: Completed at Wed Mar 22 18:01:33 UTC 2023 for expirment C96C48_hybatmDA

TerrenceMcGuinness-NOAA committed:
Update hera.sh

TerrenceMcGuinness-NOAA committed:
Update create_experiment.py

TerrenceMcGuinness-NOAA committed:
added some edification to the layout of the imput arguments in YAML input to the python bits

TerrenceMcGuinness-NOAA committed:
changed Returns to Output in main discription and input_args Returns args

TerrenceMcGuinness-NOAA committed:
spelled args correct with type and a discription fo the doc-block to the function passe_args in the creat_experiment python script

WalterKolczynski-NOAA committed:
Correct return type of input_args

The return type of `parse_args()` (which is what `input_args()` is
returning) is technically a Namespace, not a dict.
